dynamic_testing/SPL_WebServicesTesting/SPL_WebServices.md

```diff
@@ -0,0 +1,79 @@
+#Scope
+##Technical Scope
+Safe Places web services, hosted at hermes (mobile app facing) and zeus (HA facing).
+
+Used the following references
+[OWASP Testing for Session Managemnet](https://www.owasp.org/index.php/Testing_for_Session_Management).
+[OWASP JWT Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/JSON_Web_Token_for_Java_Cheat_Sheet.html)
+[JWT Attack Playbook](https://github.com/ticarpi/jwt_tool/wiki)
+[RFC 8725](https://datatracker.ietf.org/doc/rfc8725/?include_text=1)
+[jsonwebtoken docs](https://github.com/auth0/node-jsonwebtoken)
+
+## 2.10 Service Authentication Requirements
+
+* Verify that integration secrets do not rely on unchanging passwords, such as API keys or shared privileged accounts.
+
+__A username and password is provided for the authenticated APIs.  In this example spladmin/ password.  The actual authentication is just a reference LDAP implementation so out of scope of this review.  A JWT token is provided based on successful authentication.  No "hardcoded" secret is used.  This was verified on each documented API.__
+
+* Verify that if passwords are required, the credentials are not a default account.
+
+__Verified with environments team that they specifically added the user__
+
+* Verify that passwords are stored with sufficient protection to prevent offline recovery attacks, including local system access.
+
+__Not applicable, descoping__
+
+* Verify passwords, integrations with databases and third-party systems, seeds and internal secrets, and API keys are managed securely and not included in the source code or stored within source code repositories. Such storage SHOULD resist offline attacks. The use of a secure software key store (L1), hardware trusted platform module (TPM), or a hardware security module (L3) is recommended for password storage.
+
+__Verified as envirnment variables.__
+
+
+## 3.2 Session Binding Requirements
+
+* Verify the application generates a new session token on user authentication.
+
+__Checked that a new token is generated on each auth.  Checked the jsonwebtoken library is used and the version used has [no CVEs](https://www.cvedetails.com/product/61276/Auth0-Jsonwebtoken.html?vendor_id=17859)__
+
+* Verify that session tokens possess at least 64 bits of entropy.
+
+__Additionally, checked 5 sequential tokens for entropy.  I used the script in this folder called TokenEntropy.py to do this.  I discovered that 45-47 bits of entropy is present in each new token, rather than the required 64.  [Raised in Jira](https://pathcheck.atlassian.net/browse/PLACES-321)__
```

Comment:
The TokenEntropy script looks  to compute the relative entropy between pairs of tokens.
I'd think this might be a bit different than the entropy of a single token?
Indeed, when I altered the program to compare the tokens against a same length string of "0" characters it produced measurements above 140.

Reply:
The pairs of tokens are sequentially generated from the back end.  The strength of the token is directly related to it's randomness, what I am showing in here is that only ~45 bits actually change when the next token is generated, whereas 64 bits is recommended.  Does that make sense?

Reply:
I guess my question is what is the spec meaning when it says the token should have 64 bits of randomness? Two keys that were "impossible" to crack could be only 1 bit different, and in your measure would have low mutual entropy, but still be quite protective.
Suspect that having a more complex "secret" might help make the sequential entropy be higher as more entropy is being hashed into the signature via the key. If so that that would be a good thing but I'm not sure.
Just not sure your measurement technique is addressing the same entropy metric as set out in in the requirement - i.e. the "entropy of the session token"

Reply:
I read this https://owasp.org/www-community/vulnerabilities/Insufficient_Session-ID_Length - and a few other things, and inferred this requirement to be about the randomness of sequential tokens.

```
The expected number of seconds required to guess a valid session identifier is given by the equation:

Where:

B is the number of bits of entropy in the session identifier
A is the number of guesses an attacker can try each second
S is the number of valid session identifiers that are valid and available to be guessed at any given time
The number of bits of entropy in the session identifier is always less than the total number of bits in the session identifier. For example, if session identifiers were provided in ascending order, there would be close to zero bits of entropy in the session identifier no matter the identifier’s length. Assuming that the session identifiers are being generated using a good source of random numbers, we will estimate the number of bits of entropy in a session identifier to be half the total number of bits in the session identifier. For realistic identifier lengths this is possible, though perhaps optimistic.
```

Reply:
I have asked some OWASP contacts...

Reply:
I was advised that checking the algo used is sufficient, see updates!